### DISCUSSION TITLE
Projection library study

### DISCUSSION THREAD

**Discussion by sindizzy** [DotSpatial/DotSpatial#1472] [General]
So, I am working on a PR to add some fields to the ProjectionInfo class. Simple stuff to classify the type of SR. In this work I also created a test against all the projections in the built-in library DotSpatial.Projections. These are well known coordinate systems presented in proj4 format.

After my test, I see many projections with an unknown SR type. Some research show that the +proj parameter is missing in most of these projections. As a very **simple example let's look at Winkel I.** My sample below shows **Source** as how its presented in the class, **EsriPrj** is how its converted to an ESRI string, and **Proj4** is how its re-converted internally to a proj4 string. The key here is that the **Transform** class of the ProjectionInfo is left **null** due to there being [no +proj parameter](https://github.com/DotSpatial/DotSpatial/blob/51e01caca4b0b2656a63212c926c157e656585f9/Source/Core/DotSpatial.Projections/ProjectionInfo.cs#L1097).   

Because of a null Transform when we try to reproject, an exception is thrown at  DotSpatial.Projections.Reproject.ConvertToProjected  [dest.Transform.Forward](https://github.com/DotSpatial/DotSpatial/blob/63e9cc41973932e6ea06bc9a24cc6308a2841ee0/Source/Core/DotSpatial.Projections/Reproject.cs#L213)
```
World:WinkelIworld
  Source: +ellps=WGS84 +datum=WGS84 +units=m +no_defs 
  EsriPrj: PROJCS["World_Winkel_II",GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137,298.2572235629972]],PRIMEM["Greenwich",0],UNIT["Degree",0.017453292519943295]], PARAMETER["Scale_Factor",1],UNIT["Meter",1]]
  Proj4:  +datum=WGS84 +no_defs
  IsLatLon: False
  CoordinateSystemType=Unknown
  Transform=
  Datum=D_WGS_1984
  <reproject.error>

Takimng WorldSpheroid:WinkelIsphere
  Source: +a=6371000 +b=6371000 +units=m +no_defs 
  EsriPrj: PROJCS["Sphere_Winkel_II",GEOGCS["GCS_Sphere",DATUM["D_Sphere",SPHEROID["Custom",6371000,0]],PRIMEM["Greenwich",0],UNIT["Degree",0.017453292519943295]], PARAMETER["Scale_Factor",1],UNIT["Meter",1]]
  Proj4:  +a=6371000 +b=6371000 +no_defs
  IsLatLon: False
  CoordinateSystemType=Unknown
  Transform=
  Datum=D_Sphere
  <reproject.error>
```
Taking this a bit further, I go to [epsg.io](https://epsg.io/53018) and get the Winkel I proj4 strings from that source. I run those through my test harness, and these are the results:

```
ESRI:54018
  Source: +proj=wink1 +lon_0=0 +lat_ts=50.4597762521898 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs +type=crs
  EsriPrj: PROJCS["",GEOGCS["",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137,298.2572235629972]],PRIMEM["Greenwich",0],UNIT["Degree",0.017453292519943295]], PROJECTION["Winkel_I"],PARAMETER["False_Easting",0],PARAMETER["False_Northing",0],PARAMETER["Central_Meridian",0],PARAMETER["Standard_Parallel_1",50.4597762521898],PARAMETER["Scale_Factor",1],UNIT["Meter",1]]
  Proj4:  +x_0=0 +y_0=0 +lon_0=0 +lat_1=50.4597762521898 +proj=wink1 +datum=WGS84 +no_defs
  IsLatLon: False
  CoordinateSystemType=Projected
  Transform=Winkel_I
  Datum=D_WGS_1984
    Source(WGS84)                  Target(this)
  X      -96.7196               -7,942,644.0573
  Y       32.9886                3,672,274.1540
  Z        0.0000                        0.0000

ESRI:53018
  Source: +proj=wink1 +lon_0=0 +lat_ts=50.4597762521898 +x_0=0 +y_0=0 +R=6371000 +units=m +no_defs +type=crs
  EsriPrj: PROJCS["",GEOGCS["",DATUM["",SPHEROID["Custom",6371000,447.1603395551009]],PRIMEM["Greenwich",0],UNIT["Degree",0.017453292519943295]], PROJECTION["Winkel_I"],PARAMETER["False_Easting",0],PARAMETER["False_Northing",0],PARAMETER["Central_Meridian",0],PARAMETER["Standard_Parallel_1",50.4597762521898],PARAMETER["Scale_Factor",1],UNIT["Meter",1]]
  Proj4:  +x_0=0 +y_0=0 +lon_0=0 +lat_1=50.4597762521898 +proj=wink1 +a=6371000 +b=6356752.314245179 +no_defs
  IsLatLon: False
  CoordinateSystemType=Projected
  Transform=Winkel_I
  Datum=
    Source(WGS84)                  Target(this)
  X      -96.7196               -7,933,756.4071
  Y       32.9886                3,668,164.9571
  Z        0.0000                        0.0000
```

These results match up well with the [online converter ](https://epsg.io/transform#s_srs=4326&t_srs=54018&x=-96.7196000&y=32.9886000) and also with the TatukGIS calculator.

So, I propose that these projections never worked properly, and they need updating. I can attempt to do this but need some guidance on how to go about it. Currently I get 49 projections without a +proj parameter. I think most can be fixed by just updating them. 

Let me know what you think.

**Comment by jany-tenaj** [comment]
Not only the projections need updating but the whole DS.Projections project needs to be updated to the current [proj](https://proj.org/) code.

**Comment by sindizzy** [comment]
I agree. Lots of new technology since proj4 first dropped. 
I have some knowledge of projections but it would take me some time to get up to speed on all the new advances.
For now I suggest we take it in phases.
First we fix the projections in the current library.
Then we start to update to the latest Proj framework.

**Reply by jany-tenaj** [reply]
There is one fork where I tried to fix the projections and caused a lot of failing tests ;) That's why I'd say it doesn't make sense to update the projections if we don't update the proj source code as well.

**Reply by sindizzy** [reply]
I guess what I am saying is that most of these projections are not really broken. They are just missing the +proj parameter which gives them no Transform class. As I showed with the Winkel projections, you put the proper +proj in them and they work fine.

**Comment by sindizzy** [comment]
I have some time on my hands so started looking at the [proj repo](https://github.com/OSGeo/PROJ/tree/master/src/projections). That is some crazy math but I think I, at least, can follow some of it.

Looking at your [branch](https://github.com/DotSpatial/DotSpatial/tree/update-knownprojections) it appears that you were switching from FromProj4String to FromEpsgCode/FromAuthorityCode for the known projections. If I remember correctly, the EPSG lookup is a built in database that is probably sourced from an ESRI reference. I'll clone the branch and take a look. I am going to assume many were breaking because of the missing transforms in the Transform folder.

**Comment by jany-tenaj** [comment]
@sindizzy Are you able to run the projection tests? I'm getting a module not found error but have no idea why. Seems there is something missing from my machine but no idea what.

**Comment by sindizzy** [comment]
@jany-tenaj let me run the test on my end later today.

**Comment by sindizzy** [comment]
I ran the tests and it seems to imply that a null pointer is being passed somewhere. 

![image](https://user-images.githubusercontent.com/5298460/210121014-73f0c1d4-925e-45a3-a624-71cb772d7bbc.png)

![image](https://user-images.githubusercontent.com/5298460/210120973-10ae8a87-d4e9-4e10-aee1-d693fd27276c.png)

![image](https://user-images.githubusercontent.com/5298460/210120979-4afa245b-0edf-4e32-8c4d-51814c1e9155.png)

**Comment by jany-tenaj** [comment]
@sindizzy Yes that's what I saw to. The Proj4 initializer tries to load the proj.dll which seems to be c++. Without getting that fixed we can't run those tests.

**Comment by sindizzy** [comment]
@jany-tenaj did you update the Proj4 library? It seems that some things have changed.
I did a trace on the exceptions and it boils down to errors in \DotSpatial\Source\Tests\DotSpatial.Projections.Tests\Proj4.cs
In that file there is a definition for static Proj4() and that is not used anywhere else in the solution. It's only used in the Test.
I am wondering if this was left over from past branches as I now use the internal ProjectionInfo class for any projection testing.

Anyway I put some debugs in 
- static Proj4()
- public static Delegate LoadFunction<T>(string dllPath, string functionName)

with output:
Proj.dll found at \DotSpatial\Source\Tests\DotSpatial.Projections.Tests\bin\x64\proj.dll
Function: pj_init_plus in \DotSpatial\Source\Tests\DotSpatial.Projections.Tests\bin\x64\proj.dll returned a null pointer.

In summary...should this testing method be replaced by something else rather than a straight connection to the proj.dll?

**Comment by sindizzy** [comment]
More observations:
- I can't find anywhere else in the solution where proj.dll is loaded and used. Only in the Projection.Tests
- In DotSpatial\Source\Tests\DotSpatial.Projections.Tests\Projected\StatePlaneNad1927.cs there is this note: [Ignore("Tests fails only on x64. It seems proj.dll issue.")]
- Looks like some things did change with the proj4 library according to this post: https://github.com/OSGeo/PROJ/issues/2509
